pep-0577.rst

```diff
@@ -0,0 +1,652 @@
+
+Meanwhile, attributes defined in a class body are ignored for the purpose of
+defining a method's lexical closure, which means adding them there wouldn't
+work at an implementation level.
+
+Rather than trying to resolve that inherent ambiguity, this PEP simply
+prohibits such usage, and requires that any affected logic be written somewhere
+other than directly inline in the class body (e.g. in a separate helper
+function).
+
+
+Examples
+========
+
+Simplifying retry loops
+-----------------------
+
+There are currently a few different options for writing retry loops, including::
+
+    # Post-decrementing a counter
+    remaining_attempts = 9
+    while remaining_attempts:
+        try:
+            result = attempt_operation()
+        except Exception as exc:
+            remaining_attempts -= 1
+            continue # Failed, so try again
+        break # Success!
+    else:
+        # Ran out of attempts before succeeding
+        raise OperationFailed("No more attempts remaining") from exc
+
+    # Loop-and-a-half with a pre-decremented counter
+    remaining_attempts = 10
+    while True:
+        remaining_attempts -= 1
+        if not remaining_attempts:
+            # Ran out of attempts before succeeding
+            raise OperationFailed("No more attempts remaining") from exc
+        try:
+            result = attempt_operation()
+        except Exception as exc:
+            continue # Failed, so try again
+        break # Success!
+
+Each of the available options hides some aspect of the intended loop structure
+inside the loop body, whether that's the state modification, the exit condition,
+or both.
+
+The proposal in this PEP allows both the state modification and the exit
+condition to be included directly in the loop header::
+
+    remaining_attempts = 10
+    while remaining_attempts -= 1:
+        try:
+            result = attempt_operation()
+        except Exception as exc:
+            continue # Failed, so try again
+        break # Success!
+    else:
+        # Ran out of attempts before succeeding
+        raise OperationFailed("No more attempts remaining") from exc
+
+
+Simplifying if-elif chains
+--------------------------
+
+if-elif chains that need to rebind the checked condition currently need to
+be written using nested if-else statements::
+
+
+    m = pattern.match(data)
+    if m:
+        ...
+    else:
+        m = other_pattern.match(data)
+        if m:
+            ...
+        else:
+            m = yet_another_pattern.match(data)
+            if m:
+                ...
+            else:
+                ...
+
+As with PEP 572, this PEP allows the else/if portions of that chain to be
+condensed, making their consistent and mutually exclusive structure more
+readily apparent::
+
+    m = pattern.match(data)
+    if m:
+        ...
+    elif m := other_pattern.match(data):
+        ...
+    elif m := yet_another_pattern.match(data):
+        ...
+    else:
+        ...
```

Comment:
The need to write the `if` clause different from the `elif` clauses is annoying in your version though. I find myself fairly frequently rearranging the clauses of such a "poor man's match statement" and having to adjust the pattern of today's top-most check just to satisfy the arbitrary requirement on `:=` makes me unhappy.